### ISSUE TITLE
Feature: Add general purpose dead letter queue to pipeline

### ISSUE THREAD

**Issue by suyograo** [elastic/logstash#2607]
Today, when Logstash cannot process an event due to an error, it has two choices: _drop_ or _retry_. If the condition is temporary - the next stage in the pipeline is temporarily overloaded, intermittent network failues - retrying is a good approach. Otherwise, if the failure is permanent - bad encoding, Elasticsearch mapping error - retrying could cause an indefinite stall in processing by creating poison messages. In this scenario, dropping the event is preferred so we can unblock the pipeline. 

As a third option, we plan to introduce a _dead letter queue_ (DLQ), which can be used to store any event that was not successfully sent to its destination. We can also shunt events which have the potential of bringing down the pipeline. DLQ could also receive events that abuse the grok filter (e.g. runaway regular expressions which cause expensive backtracking), failures in grok patterns, date filters, and so on. 

Having a separate queue allows the administrator to examine these events and resolve problems later, while keeping the streaming nature of the pipeline. Additionally, dead letter queues also allows users to further process(clean) the events before resending to its original destination. We plan to add an input plugin that can read from the DLQ.

**Comment by suyograo** [comment]
Related issue: https://github.com/elasticsearch/logstash/issues/2533

**Comment by bitsofinfo** [comment]
What will be the implementation of the DLQ internally? Can I suggest that the impl of the DLQ be implemented by the similar mechanism driven by https://github.com/elastic/logstash/issues/2605 

In particular that it be configurable based on a abstracted "plugin" concept for a DLQ interface, logstash could provide one or two impls out of the box, but it would permit folks to dump their stuff to other DQL impls provided via plugins.

**Comment by jordansissel** [comment]
> Can I suggest that the impl of the DLQ be implemented by the similar mechanism driven by #2605

This is our thought also :)

The roadmap for new features that are possible once we have persistenc queues include a DLQ as presented in this ticket. Woo!

**Comment by jsvd** [comment]
__TL;DR:__ 

1. use queue persistence for crash handling
2. use a dead letter api for everything else.
3. the DL API can push to a DLQ or a DL pipeline "last-stage", but as a POC it can just write to a file. 4. everything else should be taken care of by a clean/safe shutdown sequence.

-----------------------

I think there's some duplication and too much revisiting of arguments regarding DLQ, persistence and "stalling".

Looking at this issue, #2605, #3697, #2867 and some internal discussions we seem to be converging on a few ideas:

#### 1. Handling poison messages

Marking an event as a `dead_letter` is a protection mechanism against poison messages, and not about environment failures (such as disconnections or slow consumers).

When either the pipeline or a plugin marks and event as a *dead letter* a few things can happen:

a) implicit: the event is "magically" injected into the `input_to_filter` queue with a special tag
b) explicit: a 4th, new, stage of the pipeline is created to receive these events and apply outputs

Option a) allows us to rerun filters and outputs against DL events but forces users to be very careful to reject DL events from their normal filters/output flow.
Option b) is a "last step" action in the pipeline, such that any event in filter and outputs sections can be sent to the `dead_letter { }` phase. This phase should be restricted to output plugins and only a few should be allowed (e.g. file, stdout)

Before implementing a) or b) we can do c): all DL events are sent to a `/tmp/dl.date` file

#### 2. Persisting internal queues for crash tolerance

This issue is about ensuring inflight events are implicitly processed after a crash or a restart. Let's look at both scenarios:

##### Logstash crash

A crash can come from a fatal failure inside of Logstash, an OS or hardware failure, or some other "unlikely event" that forces an unclean shutdown. If the buffers are persisted (e.g. @colinsurprenant's work on pluggable mmap'ed queues) and we can fingerprint the configuration then we can resume operations after an unclean shutdown.

##### Logstash restart

In common scenarios a restart should cause Logstash to do a clean shutdown such that all inputs stop processing and logstash shuts down after the inflight count goes to zero. For this there's **no need for queue persistence**.

#### 3. Handling inflight events of a stalled Logstash instance

The two main reasons for this scenario come from the two above: a poison message or a very slow (or hanged) filter/output.
This results in stalled shutdown sequence, which means a clean shutdown (without data loss) is not possible. 

#### Poison message preventing clean shutdown

Assuming we introduce the dead_letter mechanism in 1., then this will be no longer be a reason for halting the shutdown or logstash itself.

####  Slow consumers/Hanged Plugins

Potential causes:

* A permanent plugin failure: consumer died, or never existed (wrong configuration).
 * very poorly performant logstash instance, due to highly complex grok patterns, poorly written ruby filters, ridiculously slow consumers, memory leaks, gcs, etc.

In all these scenarios a restart will occur but the shutdown sequence will be stalled and therefore we need to safeguard the inflight data and force the exit.
Persisting the queues is not an option since it is likely that the configuration will change so marking all inflight data as dead_letter seems to be the logical step.

---------------------

Sorry about the long text

**Comment by colinsurprenant** [comment]
this also relates to the general exception handling strategy as discussed in #4127 and upcoming persistence work in #4767

**Comment by kurtado** [comment]
Is there a tentative milestone for this?

**Comment by colinsurprenant** [comment]
The first DLQ iteration is tracked in issue #5283

**Comment by acchen97** [comment]
Including notes here from our discussions around initial scope (outside of the PQ eviction flow):

- Pipeline DLQ (Global catch-all)
 - logstash.yml setting option
 - Opt-in
 - Plugin DLQ
- Elasticsearch output, Grok filter (could potentially extend to all filters with parse exceptions)
 - Config setting at plugin level
 - Opt-in
- DLQ config
 - Path /tmp/foobar - use PQ infra
- Add API/SIGHUP for flush/rotate to enable reingest workflow
- DLQ fills up → backpressure
- Need to create a DLQ input to close the reingest loop

**Comment by Jakauppila** [comment]
+1

We use metadata tags to send our events to a DLQ within RabbitMQ if it fails any of our processing. Basically, inject an 'unprocessed' tag at the input, determine criteria within filter and remove tag if it passes, then either send it to the DLQ or proceed to indexing in Elasticsearch.

This works great for anything we have control of, but yes, if it fails to index into Elasticsearch due to a mapping error, we have no way of handling that today.

https://gist.github.com/Jakauppila/48519e586f10c7dd81297d769be6f9ae

**Comment by talevy** [comment]
closing in favor of #6700
